File: src/launcher/MinecraftLauncher.java
package launcher;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class MinecraftLauncher {

    public interface OutputListener {
        void onLine(String line);
    }

    private final Config config;

    public MinecraftLauncher(Config config) {
        this.config = config;
    }

    /**
     * Launches the given (already installed) version for the given account.
     * versionId is the id used for the game directory/profile (e.g. "1.20.4"
     * or a fabric synthetic id like "1.20.4-fabric-0.15.11").
     */
    public Process launch(String versionId, Account account, String serverAddress, OutputListener output) throws Exception {
        Path mcDir = config.getMinecraftDir();
        Path versionDir = mcDir.resolve("versions").resolve(versionId);
        Path versionJsonFile = versionDir.resolve(versionId + ".json");
        if (!Files.exists(versionJsonFile)) {
            throw new IOException("Version metadata not found for " + versionId + ". Install it first.");
        }
        Map<String, Object> versionJson = Json.parseObject(Files.readString(versionJsonFile, StandardCharsets.UTF_8));

        // Resolve inheritance (Fabric profiles set inheritsFrom -> vanilla version)
        String inheritsFrom = Json.asString(versionJson.get("inheritsFrom"), null);
        Map<String, Object> parentJson = null;
        String parentId = null;
        if (inheritsFrom != null) {
            parentId = inheritsFrom;
            Path parentFile = mcDir.resolve("versions").resolve(parentId).resolve(parentId + ".json");
            parentJson = Json.parseObject(Files.readString(parentFile, StandardCharsets.UTF_8));
        }

        String mainClass = Json.asString(versionJson.get("mainClass"),
                parentJson != null ? Json.asString(parentJson.get("mainClass"), "net.minecraft.client.main.Main")
                        : "net.minecraft.client.main.Main");

        String jarVersionId = parentId != null ? parentId : versionId;
        Path clientJar = mcDir.resolve("versions").resolve(jarVersionId).resolve(jarVersionId + ".jar");

        // --- classpath ---
        List<String> classpath = new ArrayList<>();
        collectLibraryPaths(mcDir, versionJson, classpath);
        if (parentJson != null) collectLibraryPaths(mcDir, parentJson, classpath);
        classpath.add(clientJar.toString());

        // --- natives dir ---
        Path nativesDir = mcDir.resolve("versions").resolve(jarVersionId).resolve(jarVersionId + "-natives");

        // --- profile / game dir ---
        Path profileDir = mcDir.resolve("profiles").resolve(versionId);
        Files.createDirectories(profileDir);
        Files.createDirectories(profileDir.resolve("mods"));

        String assetIndexId = "legacy";
        Object assetIndexObj = (versionJson.get("assetIndex") != null) ? versionJson.get("assetIndex")
                : (parentJson != null ? parentJson.get("assetIndex") : null);
        if (assetIndexObj instanceof Map) {
            assetIndexId = Json.asString(Json.asObject(assetIndexObj).get("id"), "legacy");
        }

        String javaPath = config.getString("javaPath", "");
        if (javaPath == null || javaPath.isBlank()) {
            javaPath = findJavaExecutable();
        }

        int ram = config.getInt("ram", 4096);

        List<String> cmd = new ArrayList<>();
        cmd.add(javaPath);
        cmd.add("-Xmx" + ram + "M");
        cmd.add("-Xms" + Math.min(ram, 1024) + "M");
        cmd.add("-Djava.library.path=" + nativesDir.toAbsolutePath());
        String extra = config.getString("extraJvmArgs", "").trim();
        if (!extra.isEmpty()) {
            for (String tok : extra.split("\\s+")) cmd.add(tok);
        }
        cmd.add("-cp");
        cmd.add(String.join(File.pathSeparator, classpath));
        cmd.add(mainClass);

        // Game arguments (works across old and modern versions without needing
        // to parse Mojang's full conditional "arguments" schema).
        cmd.add("--username"); cmd.add(account.name);
        cmd.add("--version"); cmd.add(versionId);
        cmd.add("--gameDir"); cmd.add(profileDir.toAbsolutePath().toString());
        cmd.add("--assetsDir"); cmd.add(mcDir.resolve("assets").toAbsolutePath().toString());
        cmd.add("--assetIndex"); cmd.add(assetIndexId);
        cmd.add("--uuid"); cmd.add(account.uuid.replace("-", ""));
        cmd.add("--accessToken"); cmd.add(account.accessToken == null || account.accessToken.isBlank() ? "0" : account.accessToken);
        cmd.add("--userType"); cmd.add(account.type.equals("msa") ? "msa" : "legacy");
        cmd.add("--versionType"); cmd.add("release");

        if (serverAddress != null && !serverAddress.isBlank()) {
            String host = serverAddress;
            String port = "25565";
            if (serverAddress.contains(":")) {
                String[] parts = serverAddress.split(":", 2);
                host = parts[0];
                port = parts[1];
            }
            cmd.add("--server"); cmd.add(host);
            cmd.add("--port"); cmd.add(port);
        }

        ProcessBuilder pb = new ProcessBuilder(cmd);
        pb.directory(profileDir.toFile());
        pb.redirectErrorStream(true);
        Process process = pb.start();

        if (output != null) {
            Thread reader = new Thread(() -> {
                try (BufferedReader br = new BufferedReader(new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8))) {
                    String line;
                    while ((line = br.readLine()) != null) {
                        output.onLine(line);
                    }
                } catch (IOException ignored) {
                }
            }, "mc-output-reader");
            reader.setDaemon(true);
            reader.start();
        }

        return process;
    }

    @SuppressWarnings("unchecked")
    private void collectLibraryPaths(Path mcDir, Map<String, Object> versionJson, List<String> out) {
        Path librariesDir = mcDir.resolve("libraries");
        List<Object> libraries = Json.asArray(versionJson.get("libraries"));
        for (Object lo : libraries) {
            Map<String, Object> lib = Json.asObject(lo);
            if (!RuleUtil.isAllowed(lib)) continue;

            Map<String, Object> downloads = Json.asObject(lib.get("downloads"));
            Map<String, Object> artifact = Json.asObject(downloads.get("artifact"));
            String path = Json.asString(artifact.get("path"), null);
            if (path != null) {
                Path full = librariesDir.resolve(path);
                if (Files.exists(full)) out.add(full.toString());
                continue;
            }

            // Fabric-style {name, url} libraries have no "downloads" block
            String name = Json.asString(lib.get("name"), null);
            if (name != null) {
                String[] parts = name.split(":");
                if (parts.length >= 3) {
                    String group = parts[0].replace('.', '/');
                    String artifactId = parts[1];
                    String version = parts[2];
                    String rel = group + "/" + artifactId + "/" + version + "/" + artifactId + "-" + version + ".jar";
                    Path full = librariesDir.resolve(rel);
                    if (Files.exists(full)) out.add(full.toString());
                }
            }
        }
    }

    private String findJavaExecutable() {
        String javaHome = System.getProperty("java.home");
        String exe = System.getProperty("os.name").toLowerCase().contains("win") ? "java.exe" : "java";
        Path candidate = Path.of(javaHome, "bin", exe);
        if (Files.exists(candidate)) return candidate.toString();
        return "java"; // hope it's on PATH
    }
}

File: electron/lib/minecraftLauncher.ts
import fs from 'fs';
import path from 'path';
import { spawn, ChildProcess } from 'child_process';
import { isAllowed } from './ruleUtil';
import { coordToPath } from './mavenUtil';
import type { ConfigStore, Account } from './configStore';

export type OutputListener = (line: string) => void;

export class MinecraftLauncher {
  constructor(private config: ConfigStore) {}

  /**
   * Launches the given (already installed) version for the given account.
   * versionId is the id used for the game directory/profile (e.g. "1.20.4"
   * or a fabric synthetic id like "1.20.4-fabric-0.15.11").
   */
  async launch(
    versionId: string,
    account: Account,
    serverAddress: string | null,
    output?: OutputListener
  ): Promise<ChildProcess> {
    const mcDir = this.config.getMinecraftDir();
    const versionDir = path.join(mcDir, 'versions', versionId);
    const versionJsonFile = path.join(versionDir, `${versionId}.json`);
    if (!fs.existsSync(versionJsonFile)) {
      throw new Error(`Version metadata not found for ${versionId}. Install it first.`);
    }
    const versionJson = JSON.parse(fs.readFileSync(versionJsonFile, 'utf-8'));

    // Resolve inheritance (Fabric profiles set inheritsFrom -> vanilla version)
    const inheritsFrom: string | undefined = versionJson.inheritsFrom;
    let parentJson: any = null;
    let parentId: string | null = null;
    if (inheritsFrom) {
      parentId = inheritsFrom;
      const parentFile = path.join(mcDir, 'versions', parentId, `${parentId}.json`);
      parentJson = JSON.parse(fs.readFileSync(parentFile, 'utf-8'));
    }

    const mainClass = versionJson.mainClass ?? parentJson?.mainClass ?? 'net.minecraft.client.main.Main';

    const jarVersionId = parentId ?? versionId;
    const clientJar = path.join(mcDir, 'versions', jarVersionId, `${jarVersionId}.jar`);

    // --- classpath ---
    const classpath: string[] = [];
    this.collectLibraryPaths(mcDir, versionJson, classpath);
    if (parentJson) this.collectLibraryPaths(mcDir, parentJson, classpath);
    classpath.push(clientJar);

    // --- natives dir ---
    const nativesDir = path.join(mcDir, 'versions', jarVersionId, `${jarVersionId}-natives`);

    // --- profile / game dir ---
    const profileDir = path.join(mcDir, 'profiles', versionId);
    fs.mkdirSync(path.join(profileDir, 'mods'), { recursive: true });

    let assetIndexId = 'legacy';
    const assetIndexObj = versionJson.assetIndex ?? parentJson?.assetIndex;
    if (assetIndexObj) assetIndexId = assetIndexObj.id ?? 'legacy';

    let javaPath = this.config.data.javaPath?.trim();
    if (!javaPath) javaPath = MinecraftLauncher.resolveJavaExecutable();

    const ram = this.config.data.ram || 4096;

    const cmd: string[] = [];
    cmd.push(`-Xmx${ram}M`);
    cmd.push(`-Xms${Math.min(ram, 1024)}M`);
    cmd.push(`-Djava.library.path=${path.resolve(nativesDir)}`);
    const extra = (this.config.data.extraJvmArgs || '').trim();
    if (extra) cmd.push(...extra.split(/\s+/));
    cmd.push('-cp');
    cmd.push(classpath.join(path.delimiter));
    cmd.push(mainClass);

    // Game arguments (works across old and modern versions without needing
    // to parse Mojang's full conditional "arguments" schema).
    cmd.push('--username', account.name);
    cmd.push('--version', versionId);
    cmd.push('--gameDir', path.resolve(profileDir));
    cmd.push('--assetsDir', path.resolve(path.join(mcDir, 'assets')));
    cmd.push('--assetIndex', assetIndexId);
    cmd.push('--uuid', account.uuid);
    cmd.push('--accessToken', account.accessToken || '0');
    cmd.push('--userType', account.type === 'msa' ? 'msa' : 'legacy');
    cmd.push('--versionType', 'release');

    if (serverAddress && serverAddress.trim()) {
      let host = serverAddress;
      let port = '25565';
      if (serverAddress.includes(':')) {
        [host, port] = serverAddress.split(':', 2);
      }
      cmd.push('--server', host);
      cmd.push('--port', port);
    }

    const proc = spawn(javaPath, cmd, { cwd: profileDir });

    if (output) {
      proc.stdout.on('data', (chunk) => {
        for (const line of chunk.toString().split(/\r?\n/)) if (line) output(line);
      });
      proc.stderr.on('data', (chunk) => {
        for (const line of chunk.toString().split(/\r?\n/)) if (line) output(line);
      });
    }

    return proc;
  }

  private collectLibraryPaths(mcDir: string, versionJson: any, out: string[]) {
    const librariesDir = path.join(mcDir, 'libraries');
    const libraries: any[] = versionJson.libraries ?? [];
    for (const lib of libraries) {
      if (!isAllowed(lib)) continue;

      const artifact = lib.downloads?.artifact;
      const relPath = artifact?.path;
      if (relPath) {
        const full = path.join(librariesDir, relPath);
        if (fs.existsSync(full)) out.push(full);
        continue;
      }

      // Fabric/Forge-style {name, url} libraries have no "downloads" block
      const name = lib.name;
      if (name) {
        const rel = coordToPath(name);
        if (rel) {
          const full = path.join(librariesDir, rel);
          if (fs.existsSync(full)) out.push(full);
        }
      }
    }
  }

  static resolveJavaExecutable(configuredPath?: string | null): string {
    if (configuredPath && configuredPath.trim()) return configuredPath;
    const javaHome = process.env.JAVA_HOME;
    const exe = process.platform === 'win32' ? 'java.exe' : 'java';
    if (javaHome) {
      const candidate = path.join(javaHome, 'bin', exe);
      if (fs.existsSync(candidate)) return candidate;
    }
    return 'java';
  }
}
